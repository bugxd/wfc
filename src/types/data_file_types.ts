export interface DataTile {
  svg: string;
  north: string;
  east: string;
  south: string;
  west: string;
  rotiationDegrees: number[];
  frequency: number;
}

export interface DataFile {
  cellSize: number,
  tiles: DataTile[],
}
