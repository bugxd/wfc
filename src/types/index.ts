export interface Tile {
  svg: string;
  north: string;
  east: string;
  south: string;
  west: string;
}
export type Tiles = Tile[];

export interface Adjacent {
  tileId: number;
  north: number[];
  east: number[];
  south: number[];
  west: number[];
}

export type Adjacents = Adjacent[];

export interface EntropyCell {
  row: number;
  col: number;
  entropy: number;
}

export interface TileRemoval {
  row: number;
  col: number;
  tileId: number;
}

export enum Direction {
  NORTH = 0,
  EAST = 1,
  SOUTH = 2,
  WEST = 3,
}

export interface Neighbor {
  direction: Direction,
  row: number,
  col: number
}
export type Neighbors = Neighbor[];

export interface DataFile {
  cellSize: number,
  tiles: Tile[],
  frequencies: number[],
}

export interface TileEnablerCount  {
  tileId: number;
  north: number;
  east: number;
  south: number;
  west: number;
}

export interface TileCell {
  row: number,
  col: number,
  possible: number,
  svg: string | undefined,
}
