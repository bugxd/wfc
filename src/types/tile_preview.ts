export function getPreviewImage(cellSize: number, svg: string, north: string, east: string, south: string, west: string, degree: number): string {

  /**
   * g = group - all dynamic elements in a Group
   * transform translate - 20 20 offset to be in the middle of the text
   * transform rotate - rotate around the given center point round(cellSize/2) for given degree
   */
  return `<svg viewBox='0 0 ${cellSize+40} ${cellSize+40}' height='${cellSize+40}' width='${cellSize+40}' xmlns='http://www.w3.org/2000/svg'>
    <text x="${Math.round(cellSize/2)}" y="10" class="small">${north}</text>
    <text x="${cellSize+30}" y="${Math.round(cellSize/2)+10}" class="small" style="writing-mode: tb;">${east}</text>
    <text x="${Math.round(cellSize/2)}" y="${cellSize+30}" class="small">${south}</text>
    <text x="10" y="${Math.round(cellSize/2)+10}" class="small" style="writing-mode: tb;">${west}</text>
    <rect x="19" y="19" width="${cellSize+2}" height="${cellSize+2}" style="fill:rgb(255,255,255);stroke:rgb(255,0,0);stroke-width:1" />
    <g x='20' y='20' height='${cellSize}' width='${cellSize}' transform='translate(20, 20) rotate(${degree},${Math.round(cellSize/2)},${Math.round(cellSize/2)})'>
      ${svg}
    </g>
  </svg>`
}
