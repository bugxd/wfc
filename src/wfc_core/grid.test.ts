import Grid from "./grid";
import { Adjacents, Direction, Tiles } from "../types";


describe("grid tests", () => {
  beforeEach(() => {
    jest.spyOn(global.Math, 'random').mockReturnValue(0.234567891);
  });

  afterEach(() => {
      jest.spyOn(global.Math, 'random').mockRestore();
  });

  it("should fill grid", () => {
    const tiles: Tiles = [
      {
        svg: "+",
        north: "aba",
        east: "aba",
        south: "aba",
        west: "aba",
        rotiationDegree: 0,
        uid: "a",
      },
      {
        svg: "|",
        north: "aba",
        east: "aaa",
        south: "aba",
        west: "aaa",
        rotiationDegree: 0,
        uid: "a",
      },
      {
        svg: "-",
        north: "aaa",
        east: "aba",
        south: "aaa",
        west: "aba",
        rotiationDegree: 0,
        uid: "a",
      }
    ];
    const frequencies: number[] = [1, 1, 1];
    const adjacents: Adjacents = [
      {
        tileId: 0,
        north: [0, 1],
        east: [0, 2],
        south: [0, 1],
        west: [0, 2],
      },
      {
        tileId: 1,
        north: [0, 1],
        east: [0],
        south: [0, 1],
        west: [0],
      },
      {
        tileId: 2,
        north: [0],
        east: [0, 2],
        south: [0],
        west: [0, 2],
      }
    ];
    const sut = new Grid({
      rows: 3,
      cols: 3,
      tiles: tiles,
      frequencies: frequencies,
      adjacents: adjacents
    });

    expect(sut.rows).toBe(3);
    expect(sut.cols).toBe(3);
    expect(sut.remainingUncollapsedCells).toBe(9);

    expect(sut.grid).toMatchSnapshot();
  });

  it("should remove possible tiles", () => {
    const tiles: Tiles = [
      {
        svg: "+",
        north: "aba",
        east: "aba",
        south: "aba",
        west: "aba",
        rotiationDegree: 0,
        uid: "a",
      },
      {
        svg: "|",
        north: "aba",
        east: "aaa",
        south: "aba",
        west: "aaa",
        rotiationDegree: 0,
        uid: "a",
      },
      {
        svg: "-",
        north: "aaa",
        east: "aba",
        south: "aaa",
        west: "aba",
        rotiationDegree: 0,
        uid: "a",
      }
    ];
    const frequencies: number[] = [1, 1, 1];
    const adjacents: Adjacents = [
      {
        tileId: 0,
        north: [0, 1],
        east: [0, 2],
        south: [0, 1],
        west: [0, 2],
      },
      {
        tileId: 1,
        north: [0, 1],
        east: [0],
        south: [0, 1],
        west: [0],
      },
      {
        tileId: 2,
        north: [0],
        east: [0, 2],
        south: [0],
        west: [0, 2],
      }
    ];
    const sut = new Grid({
      rows: 3,
      cols: 3,
      tiles: tiles,
      frequencies: frequencies,
      adjacents: adjacents
    });

    sut.removePossibleTiles(1, 1, 1, 1);

    expect(sut.getCell(0, 0).possibleTiles).toStrictEqual([0, 1, 2]);
    expect(sut.getCell(1, 1).possibleTiles).toStrictEqual([0, 2]);
  });

  it("should decrease Tile Enablers By Direction", () => {
    const tiles: Tiles = [
      {
        svg: "+",
        north: "aba",
        east: "aba",
        south: "aba",
        west: "aba",
        rotiationDegree: 0,
        uid: "a",
      },
      {
        svg: "|",
        north: "aba",
        east: "aaa",
        south: "aba",
        west: "aaa",
        rotiationDegree: 0,
        uid: "a",
      },
      {
        svg: "-",
        north: "aaa",
        east: "aba",
        south: "aaa",
        west: "aba",
        rotiationDegree: 0,
        uid: "a",
      }
    ];
    const frequencies: number[] = [1, 1, 1];
    const adjacents: Adjacents = [
      {
        tileId: 0,
        north: [0, 1],
        east: [0, 2],
        south: [0, 1],
        west: [0, 2],
      },
      {
        tileId: 1,
        north: [0, 1],
        east: [0],
        south: [0, 1],
        west: [0],
      },
      {
        tileId: 2,
        north: [0],
        east: [0, 2],
        south: [0],
        west: [0, 2],
      }
    ];
    const sut = new Grid({
      rows: 3,
      cols: 3,
      tiles: tiles,
      frequencies: frequencies,
      adjacents: adjacents
    });

    sut.decreaseTileEnablersByDirection(1, 1, 1, Direction.NORTH);

    expect(sut.getCell(0, 0).tileEnablerCounts[1].north).toBe(2);
    expect(sut.getCell(1, 1).tileEnablerCounts[1].north).toBe(1);
  });

  it("should collapse a cell", () => {
    const tiles: Tiles = [
      {
        svg: "+",
        north: "aba",
        east: "aba",
        south: "aba",
        west: "aba",
        rotiationDegree: 0,
        uid: "a",
      },
      {
        svg: "|",
        north: "aba",
        east: "aaa",
        south: "aba",
        west: "aaa",
        rotiationDegree: 0,
        uid: "a",
      },
      {
        svg: "-",
        north: "aaa",
        east: "aba",
        south: "aaa",
        west: "aba",
        rotiationDegree: 0,
        uid: "a",
      }
    ];
    const frequencies: number[] = [1, 1, 1];
    const adjacents: Adjacents = [
      {
        tileId: 0,
        north: [0, 1],
        east: [0, 2],
        south: [0, 1],
        west: [0, 2],
      },
      {
        tileId: 1,
        north: [0, 1],
        east: [0],
        south: [0, 1],
        west: [0],
      },
      {
        tileId: 2,
        north: [0],
        east: [0, 2],
        south: [0],
        west: [0, 2],
      }
    ];
    const sut = new Grid({
      rows: 3,
      cols: 3,
      tiles: tiles,
      frequencies: frequencies,
      adjacents: adjacents
    });

    const result = sut.collapseCell(1, 1);

    expect(result).toStrictEqual([1, 2]);
    expect(sut.grid).toMatchSnapshot();
  });
})
