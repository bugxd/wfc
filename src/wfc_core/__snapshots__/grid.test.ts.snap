// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`grid tests should collapse a cell 1`] = `
Array [
  Array [
    Cell {
      "col": 0,
      "collapsed": false,
      "entropyNois": 0.000007654321090000001,
      "possibleTiles": Array [
        0,
        1,
        2,
      ],
      "row": 0,
      "selectedTileId": -1,
      "sumPossibleTileWeighsLogWeights": 0,
      "sumPossibleTileWeights": 3,
      "tileEnablerCounts": Array [
        Object {
          "east": 2,
          "north": 2,
          "south": 2,
          "tileId": 0,
          "west": 2,
        },
        Object {
          "east": 1,
          "north": 2,
          "south": 2,
          "tileId": 1,
          "west": 1,
        },
        Object {
          "east": 2,
          "north": 1,
          "south": 1,
          "tileId": 2,
          "west": 2,
        },
      ],
    },
    Cell {
      "col": 1,
      "collapsed": false,
      "entropyNois": 0.000007654321090000001,
      "possibleTiles": Array [
        0,
        1,
        2,
      ],
      "row": 0,
      "selectedTileId": -1,
      "sumPossibleTileWeighsLogWeights": 0,
      "sumPossibleTileWeights": 3,
      "tileEnablerCounts": Array [
        Object {
          "east": 2,
          "north": 2,
          "south": 2,
          "tileId": 0,
          "west": 2,
        },
        Object {
          "east": 1,
          "north": 2,
          "south": 2,
          "tileId": 1,
          "west": 1,
        },
        Object {
          "east": 2,
          "north": 1,
          "south": 1,
          "tileId": 2,
          "west": 2,
        },
      ],
    },
    Cell {
      "col": 2,
      "collapsed": false,
      "entropyNois": 0.000007654321090000001,
      "possibleTiles": Array [
        0,
        1,
        2,
      ],
      "row": 0,
      "selectedTileId": -1,
      "sumPossibleTileWeighsLogWeights": 0,
      "sumPossibleTileWeights": 3,
      "tileEnablerCounts": Array [
        Object {
          "east": 2,
          "north": 2,
          "south": 2,
          "tileId": 0,
          "west": 2,
        },
        Object {
          "east": 1,
          "north": 2,
          "south": 2,
          "tileId": 1,
          "west": 1,
        },
        Object {
          "east": 2,
          "north": 1,
          "south": 1,
          "tileId": 2,
          "west": 2,
        },
      ],
    },
  ],
  Array [
    Cell {
      "col": 0,
      "collapsed": false,
      "entropyNois": 0.000007654321090000001,
      "possibleTiles": Array [
        0,
        1,
        2,
      ],
      "row": 1,
      "selectedTileId": -1,
      "sumPossibleTileWeighsLogWeights": 0,
      "sumPossibleTileWeights": 3,
      "tileEnablerCounts": Array [
        Object {
          "east": 2,
          "north": 2,
          "south": 2,
          "tileId": 0,
          "west": 2,
        },
        Object {
          "east": 1,
          "north": 2,
          "south": 2,
          "tileId": 1,
          "west": 1,
        },
        Object {
          "east": 2,
          "north": 1,
          "south": 1,
          "tileId": 2,
          "west": 2,
        },
      ],
    },
    Cell {
      "col": 1,
      "collapsed": true,
      "entropyNois": 0.000007654321090000001,
      "possibleTiles": Array [],
      "row": 1,
      "selectedTileId": 0,
      "sumPossibleTileWeighsLogWeights": 0,
      "sumPossibleTileWeights": 3,
      "tileEnablerCounts": Array [
        Object {
          "east": 2,
          "north": 2,
          "south": 2,
          "tileId": 0,
          "west": 2,
        },
        Object {
          "east": 1,
          "north": 2,
          "south": 2,
          "tileId": 1,
          "west": 1,
        },
        Object {
          "east": 2,
          "north": 1,
          "south": 1,
          "tileId": 2,
          "west": 2,
        },
      ],
    },
    Cell {
      "col": 2,
      "collapsed": false,
      "entropyNois": 0.000007654321090000001,
      "possibleTiles": Array [
        0,
        1,
        2,
      ],
      "row": 1,
      "selectedTileId": -1,
      "sumPossibleTileWeighsLogWeights": 0,
      "sumPossibleTileWeights": 3,
      "tileEnablerCounts": Array [
        Object {
          "east": 2,
          "north": 2,
          "south": 2,
          "tileId": 0,
          "west": 2,
        },
        Object {
          "east": 1,
          "north": 2,
          "south": 2,
          "tileId": 1,
          "west": 1,
        },
        Object {
          "east": 2,
          "north": 1,
          "south": 1,
          "tileId": 2,
          "west": 2,
        },
      ],
    },
  ],
  Array [
    Cell {
      "col": 0,
      "collapsed": false,
      "entropyNois": 0.000007654321090000001,
      "possibleTiles": Array [
        0,
        1,
        2,
      ],
      "row": 2,
      "selectedTileId": -1,
      "sumPossibleTileWeighsLogWeights": 0,
      "sumPossibleTileWeights": 3,
      "tileEnablerCounts": Array [
        Object {
          "east": 2,
          "north": 2,
          "south": 2,
          "tileId": 0,
          "west": 2,
        },
        Object {
          "east": 1,
          "north": 2,
          "south": 2,
          "tileId": 1,
          "west": 1,
        },
        Object {
          "east": 2,
          "north": 1,
          "south": 1,
          "tileId": 2,
          "west": 2,
        },
      ],
    },
    Cell {
      "col": 1,
      "collapsed": false,
      "entropyNois": 0.000007654321090000001,
      "possibleTiles": Array [
        0,
        1,
        2,
      ],
      "row": 2,
      "selectedTileId": -1,
      "sumPossibleTileWeighsLogWeights": 0,
      "sumPossibleTileWeights": 3,
      "tileEnablerCounts": Array [
        Object {
          "east": 2,
          "north": 2,
          "south": 2,
          "tileId": 0,
          "west": 2,
        },
        Object {
          "east": 1,
          "north": 2,
          "south": 2,
          "tileId": 1,
          "west": 1,
        },
        Object {
          "east": 2,
          "north": 1,
          "south": 1,
          "tileId": 2,
          "west": 2,
        },
      ],
    },
    Cell {
      "col": 2,
      "collapsed": false,
      "entropyNois": 0.000007654321090000001,
      "possibleTiles": Array [
        0,
        1,
        2,
      ],
      "row": 2,
      "selectedTileId": -1,
      "sumPossibleTileWeighsLogWeights": 0,
      "sumPossibleTileWeights": 3,
      "tileEnablerCounts": Array [
        Object {
          "east": 2,
          "north": 2,
          "south": 2,
          "tileId": 0,
          "west": 2,
        },
        Object {
          "east": 1,
          "north": 2,
          "south": 2,
          "tileId": 1,
          "west": 1,
        },
        Object {
          "east": 2,
          "north": 1,
          "south": 1,
          "tileId": 2,
          "west": 2,
        },
      ],
    },
  ],
]
`;

exports[`grid tests should fill grid 1`] = `
Array [
  Array [
    Cell {
      "col": 0,
      "collapsed": false,
      "entropyNois": 0.000007654321090000001,
      "possibleTiles": Array [
        0,
        1,
        2,
      ],
      "row": 0,
      "selectedTileId": -1,
      "sumPossibleTileWeighsLogWeights": 0,
      "sumPossibleTileWeights": 3,
      "tileEnablerCounts": Array [
        Object {
          "east": 2,
          "north": 2,
          "south": 2,
          "tileId": 0,
          "west": 2,
        },
        Object {
          "east": 1,
          "north": 2,
          "south": 2,
          "tileId": 1,
          "west": 1,
        },
        Object {
          "east": 2,
          "north": 1,
          "south": 1,
          "tileId": 2,
          "west": 2,
        },
      ],
    },
    Cell {
      "col": 1,
      "collapsed": false,
      "entropyNois": 0.000007654321090000001,
      "possibleTiles": Array [
        0,
        1,
        2,
      ],
      "row": 0,
      "selectedTileId": -1,
      "sumPossibleTileWeighsLogWeights": 0,
      "sumPossibleTileWeights": 3,
      "tileEnablerCounts": Array [
        Object {
          "east": 2,
          "north": 2,
          "south": 2,
          "tileId": 0,
          "west": 2,
        },
        Object {
          "east": 1,
          "north": 2,
          "south": 2,
          "tileId": 1,
          "west": 1,
        },
        Object {
          "east": 2,
          "north": 1,
          "south": 1,
          "tileId": 2,
          "west": 2,
        },
      ],
    },
    Cell {
      "col": 2,
      "collapsed": false,
      "entropyNois": 0.000007654321090000001,
      "possibleTiles": Array [
        0,
        1,
        2,
      ],
      "row": 0,
      "selectedTileId": -1,
      "sumPossibleTileWeighsLogWeights": 0,
      "sumPossibleTileWeights": 3,
      "tileEnablerCounts": Array [
        Object {
          "east": 2,
          "north": 2,
          "south": 2,
          "tileId": 0,
          "west": 2,
        },
        Object {
          "east": 1,
          "north": 2,
          "south": 2,
          "tileId": 1,
          "west": 1,
        },
        Object {
          "east": 2,
          "north": 1,
          "south": 1,
          "tileId": 2,
          "west": 2,
        },
      ],
    },
  ],
  Array [
    Cell {
      "col": 0,
      "collapsed": false,
      "entropyNois": 0.000007654321090000001,
      "possibleTiles": Array [
        0,
        1,
        2,
      ],
      "row": 1,
      "selectedTileId": -1,
      "sumPossibleTileWeighsLogWeights": 0,
      "sumPossibleTileWeights": 3,
      "tileEnablerCounts": Array [
        Object {
          "east": 2,
          "north": 2,
          "south": 2,
          "tileId": 0,
          "west": 2,
        },
        Object {
          "east": 1,
          "north": 2,
          "south": 2,
          "tileId": 1,
          "west": 1,
        },
        Object {
          "east": 2,
          "north": 1,
          "south": 1,
          "tileId": 2,
          "west": 2,
        },
      ],
    },
    Cell {
      "col": 1,
      "collapsed": false,
      "entropyNois": 0.000007654321090000001,
      "possibleTiles": Array [
        0,
        1,
        2,
      ],
      "row": 1,
      "selectedTileId": -1,
      "sumPossibleTileWeighsLogWeights": 0,
      "sumPossibleTileWeights": 3,
      "tileEnablerCounts": Array [
        Object {
          "east": 2,
          "north": 2,
          "south": 2,
          "tileId": 0,
          "west": 2,
        },
        Object {
          "east": 1,
          "north": 2,
          "south": 2,
          "tileId": 1,
          "west": 1,
        },
        Object {
          "east": 2,
          "north": 1,
          "south": 1,
          "tileId": 2,
          "west": 2,
        },
      ],
    },
    Cell {
      "col": 2,
      "collapsed": false,
      "entropyNois": 0.000007654321090000001,
      "possibleTiles": Array [
        0,
        1,
        2,
      ],
      "row": 1,
      "selectedTileId": -1,
      "sumPossibleTileWeighsLogWeights": 0,
      "sumPossibleTileWeights": 3,
      "tileEnablerCounts": Array [
        Object {
          "east": 2,
          "north": 2,
          "south": 2,
          "tileId": 0,
          "west": 2,
        },
        Object {
          "east": 1,
          "north": 2,
          "south": 2,
          "tileId": 1,
          "west": 1,
        },
        Object {
          "east": 2,
          "north": 1,
          "south": 1,
          "tileId": 2,
          "west": 2,
        },
      ],
    },
  ],
  Array [
    Cell {
      "col": 0,
      "collapsed": false,
      "entropyNois": 0.000007654321090000001,
      "possibleTiles": Array [
        0,
        1,
        2,
      ],
      "row": 2,
      "selectedTileId": -1,
      "sumPossibleTileWeighsLogWeights": 0,
      "sumPossibleTileWeights": 3,
      "tileEnablerCounts": Array [
        Object {
          "east": 2,
          "north": 2,
          "south": 2,
          "tileId": 0,
          "west": 2,
        },
        Object {
          "east": 1,
          "north": 2,
          "south": 2,
          "tileId": 1,
          "west": 1,
        },
        Object {
          "east": 2,
          "north": 1,
          "south": 1,
          "tileId": 2,
          "west": 2,
        },
      ],
    },
    Cell {
      "col": 1,
      "collapsed": false,
      "entropyNois": 0.000007654321090000001,
      "possibleTiles": Array [
        0,
        1,
        2,
      ],
      "row": 2,
      "selectedTileId": -1,
      "sumPossibleTileWeighsLogWeights": 0,
      "sumPossibleTileWeights": 3,
      "tileEnablerCounts": Array [
        Object {
          "east": 2,
          "north": 2,
          "south": 2,
          "tileId": 0,
          "west": 2,
        },
        Object {
          "east": 1,
          "north": 2,
          "south": 2,
          "tileId": 1,
          "west": 1,
        },
        Object {
          "east": 2,
          "north": 1,
          "south": 1,
          "tileId": 2,
          "west": 2,
        },
      ],
    },
    Cell {
      "col": 2,
      "collapsed": false,
      "entropyNois": 0.000007654321090000001,
      "possibleTiles": Array [
        0,
        1,
        2,
      ],
      "row": 2,
      "selectedTileId": -1,
      "sumPossibleTileWeighsLogWeights": 0,
      "sumPossibleTileWeights": 3,
      "tileEnablerCounts": Array [
        Object {
          "east": 2,
          "north": 2,
          "south": 2,
          "tileId": 0,
          "west": 2,
        },
        Object {
          "east": 1,
          "north": 2,
          "south": 2,
          "tileId": 1,
          "west": 1,
        },
        Object {
          "east": 2,
          "north": 1,
          "south": 1,
          "tileId": 2,
          "west": 2,
        },
      ],
    },
  ],
]
`;
